### PR TITLE
flow: fix dead lock while stopping using callback instead of lock

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun tests

**Comment by safchain** [comment]
rerun skydive-coverage

**Comment by safchain** [comment]
rerun skydive-coverage

**Comment by safchain** [comment]
rerun skydive-coverage

**Comment by safchain** [comment]
rerun tests

File: flow/ondemand/server/server.go [hunk 1]
```diff
@@ -184,48 +184,60 @@ func (o *OnDemandProbeServer) OnWSMessage(c shttp.WSClient, msg shttp.WSMessage)
 
 	status := http.StatusBadRequest
 
-	o.Graph.Lock()
-
 	switch msg.Type {
 	case "CaptureStart":
+		o.Graph.Lock()
```
Comment (nplanel) [inline flow/ondemand/server/server.go]: why not leaving the o.Graph.Lock() around the switch

**Comment by safchain** [comment]
rerun tests

File: flow/ondemand/server/server.go [hunk 2]
```diff
@@ -214,7 +229,7 @@ func (o *OnDemandProbeServer) OnWSMessage(c shttp.WSClient, msg shttp.WSMessage)
 
 		status = http.StatusOK
 		if ok := o.unregisterProbe(n); ok {
-			o.Graph.DelMetadata(n, "Capture")
+			//o.Graph.DelMetadata(n, "Capture")
```
Comment (nplanel) [inline flow/ondemand/server/server.go]: ?

File: flow/probes/ovssflow.go [hunk 3]
```diff
@@ -299,13 +301,16 @@ func (o *OvsSFlowProbesHandler) unregisterProbe(bridgeUUID string) error {
...
-func (o *OvsSFlowProbesHandler) UnregisterProbe(n *graph.Node) error {
+func (o *OvsSFlowProbesHandler) UnregisterProbe(n *graph.Node, e FlowProbeEventHandler) error {
 	if isOvsBridge(n) {
 		if uuid, _ := n.GetFieldString("UUID"); uuid != "" {
 			err := o.unregisterProbe(uuid)
 			if err != nil {
 				return err
 			}
+			if e != nil {
+				go e.OnStopped()
```
Comment (nplanel) [inline flow/probes/ovssflow.go]: we should notify OnStopped() even if unregisterProbe() failed

File: flow/probes/probes.go [hunk 4]
```diff
@@ -86,7 +91,7 @@ func (fpb *FlowProbeBundle) UnregisterAllProbes() {
 	for _, n := range fpb.Graph.GetNodes(graph.Metadata{}) {
 		for _, p := range fpb.ProbeBundle.Probes {
 			fprobe := p.(*FlowProbe)
-			fprobe.UnregisterProbe(n)
+			fprobe.UnregisterProbe(n, nil)
```
Comment (nplanel) [inline flow/probes/probes.go]: why not storing FlowProbeEventHandler in the probe to use it here

**Pull request merged**